commander3/src/comm_nonlin_mod.f90

```diff
@@ -3306,22 +4082,107 @@ function comp_lnL_marginal_diagonal(mixing,invN_arr,data,use_det,arr_len)
     comp_lnL_marginal_diagonal = 0.d0
 
     if (MNM /= 0.d0) then 
-       MNM=1.d0/MNM !invert 1x1 matrix
+       invMNM=1.d0/MNM !invert 1x1 matrix
     else
-       comp_lnL_marginal_diagonal=1.d30 !MNM = 0.d0, i.e. no inversion possible 
+       comp_lnL_marginal_diagonal=-1.d30 !MNM = 0.d0, i.e. no inversion possible 
        deallocate(MN)
        return
     end if
 
-    comp_lnL_marginal_diagonal = 0.5d0*MNd*MNM*MNd
+    comp_lnL_marginal_diagonal = 0.5d0*MNd*invMNM*MNd
 
     !determinant of 1x1 matrix is the value of the matrix itself
-    if (use_det) comp_lnL_marginal_diagonal = comp_lnL_marginal_diagonal - 0.5d0*log(MNM) 
+    if (use_det) comp_lnL_marginal_diagonal = comp_lnL_marginal_diagonal - 0.5d0*log(invMNM) 
 
     deallocate(MN)
+  end function comp_lnL_marginal_diagonal
 
+  function comp_lnL_max_chisq_diagonal(mixing,invN_arr,data,use_det,arr_len)
+    implicit none
+    logical(lgt),               intent(in)           :: use_det
+    real(dp),     dimension(:), intent(in)           :: mixing
+    real(dp),     dimension(:), intent(in)           :: invN_arr
+    real(dp),     dimension(:), intent(in)           :: data
+    integer(i4b),               intent(in), optional :: arr_len
+    real(dp)                                         :: comp_lnL_max_chisq_diagonal
 
-  end function comp_lnL_marginal_diagonal
```

Comment:
maybe we could add a comm_nonlin_utils.f90 that holds all these math helper functions just because nonli_mod is already so long

Reply:
Agreed